### PR TITLE
[Arc] Add Comb to MLIR Arith conversion pass

### PR DESCRIPTION
**Pull request by fabianschuiki**
Add an experimental pass to convert operations in the Comb dialect to their equivalent in the Arith dialect. The pass isn't used anywhere yet, but conversion to Arith allows all canonicalizations defined on that dialect to take effect before mapping things to LLVM to generate a software model.

### CODE REVIEW

File: lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
+//===- ConvertCombToArith.cpp ---------------------------------------------===//
```
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp:1]: Should this be moved to the `Conversions` directory as it is a clean dialect-to-dialect conversion pass?

File: lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
```
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp:12]: Unused.

File: lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  using OpConversionPattern<ICmpOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ICmpOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    CmpIPredicate pred;
+    switch (adaptor.getPredicate()) {
+    case ICmpPredicate::cne:
+    case ICmpPredicate::wne:
```
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp:73]: Not sure if we should convert the 4-state predicates implicitly or exit with an error. But all other comb ops are also just lowered to 2-state even if they might contain 4-state, so I guess it's fine for now.

**Review by maerhart — APPROVED** [review]

File: lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp (hunk 1 above)
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp:1]: Yeah great point. Let me do that.

File: lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp (hunk 3 above)
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp:73]: Totally agree, this is something we'll have to tackle at some point. We might want to have a pass that lowers 4-state ops into 2-state ops with additional separate ops like `hw.is_x` or `hw.inject_x` to implement the X propagation on a side channel 🤔

**Comment by mortbopet** [comment]
Would this also subsume the [CombToLLVM](https://github.com/llvm/circt/blob/b5a984ac3a39596caeed00286db6c78a8a5f0d93/lib/Conversion/CombToLLVM/CombToLLVM.cpp) pass? That is, can we remove CombToLLVM and just rely on Comb -> Arith -> LLVM?

Nit: `arc` namespace and prefixed functions should probably be removed as well, when this is moved into the `Conversion` dir.

**Comment by maerhart** [comment]
> Would this also subsume the [CombToLLVM](https://github.com/llvm/circt/blob/b5a984ac3a39596caeed00286db6c78a8a5f0d93/lib/Conversion/CombToLLVM/CombToLLVM.cpp) pass?

Yes, that's the idea. That way we can use the vector support of the arith operations and also provide a more fine-grained lowering to those that use CIRCT as a library. The only (little) issue in completely removing CombToLLVM is the `comb.parity` operation, for which no equivalent in arith exists (at least the last time I checked). I have to see what we do with that one.

**Comment by fabianschuiki** [comment]
Do we actually use `comb.parity` anywhere? If yes, I wonder why we have a xor-reduce op called parity, but now and-reduce / or-reduce, which would probably be equally useful 🤔 Maybe we should have those as well and express them in arith through icmp? But @maerhart's point is still valid -- there's no nice xor reduction op in arith.

**Comment by maerhart** [comment]
> Do we actually use `comb.parity` anywhere?

There is a canonicalization from `comb.xor`, I don't know if any of the frontends lower directly to it. I have not encountered it in the designs I usually play with, but that might be different for others.

For AND and OR reduction we use the icmp operation by comparing to -1 and 0 respectively (with the same rationale as we don't have a NOT operation). If I remember correctly, we even had separate reduction operations back when HW was still called RTL, but removed them again.

**Comment by fabianschuiki** [comment]
Since it's not totally clear yet how to handle `comb.parity`, I'll just land this as a CombToArith conversion and then we can figure out the details of replacing CombToLLVM with this pass plus the usual ArithToLLVM lowering.

**Comment by darthscsi** [comment]
> 

`comb.parity` is generated from firrtl.xorr which is generated in real designs and is part of the firrtl spec.

File: lib/Conversion/CombToArith/CombToArith.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+/// Lower a comb::ICmpOp operation to a arith::CmpIOp
+struct IcmpOpConversion : OpConversionPattern<ICmpOp> {
+  using OpConversionPattern<ICmpOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ICmpOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    CmpIPredicate pred;
+    switch (adaptor.getPredicate()) {
```
Comment (darthscsi) [inline lib/Conversion/CombToArith/CombToArith.cpp:71]: Could also be a table lookup.  Probably more clear this way.

**Review by darthscsi — APPROVED** [review]
LGTM.

**Pull request merged**